client/src/main/scala/com/dwolla/cloudflare/RateLimitClient.scala

```diff
@@ -42,68 +42,41 @@ object RateLimitClient {
   val uriRegex: Regex = """https://api.cloudflare.com/client/v4/zones/(.+?)/rate_limits/(.+)""".r
 }
 
-object RateLimitClientImpl {
-  val notFoundCodes = List(10001)
-}
-
 class RateLimitClientImpl[F[_] : Sync](executor: StreamingCloudflareApiExecutor[F]) extends RateLimitClient[F] with Http4sClientDsl[F] {
-  import RateLimitClientImpl._
-
-  implicit val rateLimitActionDtoDecoder: Decoder[RateLimitActionDTO] =
-    (c: HCursor) ⇒ for {
-      mode ← c.downField("mode").as[String]
-      timeout ← c.downField("timeout").as[Option[Int]]
-      response ← c.downField("response").as[Option[RateLimitActionResponseDTO]]
-      dto ← (mode, timeout) match {
-        case ("challenge", None) ⇒ Right(ChallengeRateLimitActionDTO)
-        case ("js_challenge", None) ⇒ Right(JsChallengeRateLimitActionDTO)
-        case ("ban", Some(t)) ⇒ Right(BanRateLimitActionDTO(t, response))
-        case ("simulate", Some(t)) ⇒ Right(SimulateRateLimitActionDTO(t, response))
-        case _ ⇒ Left(DecodingFailure("Invalid mode for rate limit action", c.history))
-      }
-    } yield dto
-
-  override def list(zoneId: ZoneId): Stream[F, RateLimit] = {
+  private def fetch(reqF: F[Request[F]]): Stream[F, RateLimit] =
     for {
-      req ← Stream.eval(GET(BaseUrl / "zones" / zoneId / "rate_limits"))
-      record ← executor.fetch[RateLimitDTO](req)
-    } yield record
-  }
+      req ← Stream.eval(reqF)
+      res ← executor.fetch[RateLimit](req)
+    } yield res
+
+  override def list(zoneId: ZoneId): Stream[F, RateLimit] =
+    fetch(GET(BaseUrl / "zones" / zoneId / "rate_limits"))
 
   override def getById(zoneId: ZoneId, rateLimitId: String): Stream[F, RateLimit] =
-    for {
-      req ← Stream.eval(GET(BaseUrl / "zones" / zoneId / "rate_limits" / rateLimitId))
-      res ← executor.fetch[RateLimitDTO](req).returningEmptyOnErrorCodes(notFoundCodes: _*)
-    } yield res
+    fetch(GET(BaseUrl / "zones" / zoneId / "rate_limits" / rateLimitId))
 
-  override def create(zoneId: ZoneId, rateLimit: CreateRateLimit): Stream[F, RateLimit] = {
-    for {
-      req ← Stream.eval(POST(BaseUrl / "zones" / zoneId / "rate_limits", rateLimit.asJson))
-      resp ← executor.fetch[RateLimitDTO](req).map(Implicits.toModel)
-    } yield resp
-  }
+  override def create(zoneId: ZoneId, rateLimit: RateLimit): Stream[F, RateLimit] =
+    fetch(POST(BaseUrl / "zones" / zoneId / "rate_limits", rateLimit.asJson))
 
-  override def update(zoneId: ZoneId, rateLimit: RateLimit): Stream[F, RateLimit] = {
-    for {
-      req ← Stream.eval(PUT(BaseUrl / "zones" / zoneId / "rate_limits" / rateLimit.id, toDto(rateLimit).asJson))
-      resp ← executor.fetch[RateLimitDTO](req).map(Implicits.toModel)
-    } yield resp
-  }
+  override def update(zoneId: ZoneId, rateLimit: RateLimit): Stream[F, RateLimit] =
+  // TODO it would really be better to do this check at compile time by baking the identification question into the types
```

Comment:
Wasn't this TODO satisfied by the old definition of `RateLimit` that didn't have an optional ID?

Reply:
As per the API documentation (see https://github.com/Dwolla/scala-cloudflare/pull/31#discussion_r325918278), the RateLimit ID is optional, so this comment should still be valid.